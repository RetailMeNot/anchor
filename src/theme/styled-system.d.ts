/* tslint:disable no-import-side-effect*/
import 'styled-system';

// this is temporary until @types/styled-system is updated for styled-system v5
declare module 'styled-system' {
    export function system(args?: {}): {
        (props: any): {};
        config: any;
        propNames: string[];
        cache: {};
    };

    export function get(obj: any, key: string | number, fallback: any): any;
}
