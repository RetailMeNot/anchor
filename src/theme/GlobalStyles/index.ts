export { GlobalCSS } from './Globals.styles';
export { NormalizeCSS } from './Normalize.styles';
