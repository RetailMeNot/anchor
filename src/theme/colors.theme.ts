export default {
  neutrals: {
    charcoal: {
      base: '#323232',
      light: '#595959',
      dark: '#000000',
    },
    ash: {
      base: '#A6A6A6',
      light: '#D3D3D3',
      dark: '#808080',
    },
    silver: {
      base: '#F1F1F1',
      light: '#FAFAFA',
      dark: '#E7E7E7',
    },
    lavenderWhispers: {
      base: '#F7F5FC',
      light: '#FBFAFF',
      dark: '#F1EEFA',
    },
    white: {
      base: '#FFFFFF',
    },
  },
  main: {
    grapePurchase: {
      base: '#602D6C',
      light: '#91609C',
      dark: '#4B2354',
    },
    rx: {
      base: '#F23466',
      light: '#FF6B9C',
      dark: '#CC2D52',
    },
    genie: {
      base: '#4C18F3',
      light: '#7855D9',
      dark: '#42199A',
    },
  },
  accent: {
    savvyCyan: {
      base: '#0998D6',
      light: '#66CCFF',
      dark: '#0074A6',
    },
    tealBreaker: {
      base: '#00CCCD',
      light: '#00E0D1',
      dark: '#00BDBD',
    },
  },
  secondary: {
    dealEnvy: {
      base: '#14A382',
      light: '#00CDAD',
      dark: '#268068',
    },
    flashPink: {
      base: '#DF206C',
      light: '#E665A1',
      dark: '#B81D5B',
    },
    cyberMango: {
      base: '#EB663D',
      light: '#FA9F5A',
      dark: '#C74E30',
    },
    goldMine: {
      base: '#F5A623',
      light: '#FFC761',
      dark: '#CC861D',
    },
    fireSale: {
      base: '#EA4545',
      light: '#F16667',
      dark: '#CC3B3B',
    },
  },
  alert: {
    error: {
      base: '#CC3B3B',
    },
    success: {
      base: '#268068',
    },
  },
  rx: {
    purple: {
      base: '#A678B5',
    },
  },
};
