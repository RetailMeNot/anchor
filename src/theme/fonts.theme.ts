// VARIABLES
import variables from './variables.theme';

const { baseFontSize, fontFamily } = variables;

export default {
  baseFontSize,
  fontFamily,
  paragraph: {
    base: {
      fontSize: `${baseFontSize}px`,
      lineHeight: `${baseFontSize * 1.5}px`,
      fontWeight: 'normal',
    },
    informational: {
      fontSize: `${baseFontSize *  0.875}px`,
      lineHeight: `${baseFontSize * 1.125}px`,
      fontWeight: 'normal',
    },
  },
  navigation: {
    base: {
      fontSize: `${baseFontSize * .75}px`,
      lineHeight: `${baseFontSize}px`,
      fontWeight: '500',
    },
    menu: {
      fontSize: `${baseFontSize * .875}px`,
      lineHeight: `${baseFontSize}px`,
      small: {
        base: {
          fontSize: `${baseFontSize * 0.75}px`,
          fontWeight: '500',
        },
        active: {
          fontWeight: '700',
        },
      },
      large: {
        base: {
          fontWeight: '600',
        },
        active: {
          fontWeight: '700',
        },
      },
    },
  },
  heading: {
    h4: {
      fontSize: `${baseFontSize * 1.5}px`,
      lineHeight: `${baseFontSize * 2}px`,
      fontWeight: 'normal',
    },
    h3: {
      fontSize: `${baseFontSize * 1.75}px`,
      lineHeight: `${baseFontSize * 2}px`,
      fontWeight: 'normal',
    },
    h2: {
      fontSize: `${baseFontSize * 2}px`,
      lineHeight: `${baseFontSize * 2.5}px`,
      fontWeight: 'normal',
    },
    h1: {
      fontSize: `${baseFontSize * 2.25}px`,
      lineHeight: `${baseFontSize * 2.5}px`,
      fontWeight: 'normal',
    },
  },
  typography: {
    caption: {
      fontSize: `${baseFontSize * 0.75}px`,
      lineHeight: `${baseFontSize}px`,
      fontWeight: '500',
    },
    ui: {
      fontSize: `${baseFontSize * 0.875}px`,
      lineHeight: `${baseFontSize * 1.125}px`,
      fontWeight: 'normal',
    },
    body: {
      fontSize: `${baseFontSize}px`,
      lineHeight: `${baseFontSize * 1.5}px`,
      fontWeight: 'normal',
    },
    subHeading: {
      fontSize: `${baseFontSize * 1.125}px`,
      lineHeight: `${baseFontSize * 1.5}px`,
      fontWeight: 'normal',
    },
    subHeadingLarge: {
      fontSize: `${baseFontSize * 1.25}px`,
      lineHeight: `${baseFontSize * 1.5}px`,
      fontWeight: 'normal',
    },
    header: {
      small: {
        fontSize: `${baseFontSize * 1.5}px`,
        lineHeight: `${baseFontSize * 2}px`,
        fontWeight: 'normal',
      },
      medium: {
        fontSize: `${baseFontSize * 1.75}px`,
        lineHeight: `${baseFontSize * 2}px`,
        fontWeight: 'normal',
      },
      large: {
        fontSize: `${baseFontSize * 2}px`,
        lineHeight: `${baseFontSize * 2.5}px`,
        fontWeight: 'normal',
      },
      extraLarge: {
        fontSize: `${baseFontSize * 2.25}px`,
        lineHeight: `${baseFontSize * 2.5}px`,
        fontWeight: 'normal',
      },
    },
    display: {
      small: {
        fontSize: `${baseFontSize * 2.5}px`,
        lineHeight: `${baseFontSize * 3}px`,
        fontWeight: 'normal',
      },
      medium: {
        fontSize: `${baseFontSize * 3}px`,
        lineHeight: `${baseFontSize * 3.25}px`,
        fontWeight: 'normal',
      },
      large: {
        fontSize: `${baseFontSize * 3.5}px`,
        lineHeight: `${baseFontSize * 4}px`,
        fontWeight: 'normal',
      },
      extraLarge: {
        fontSize: `${baseFontSize * 4}px`,
        lineHeight: `${baseFontSize * 4.5}px`,
        fontWeight: 'normal',
      },
    },
  },
};
