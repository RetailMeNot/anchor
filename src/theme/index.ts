export { default as colors } from './colors.theme';
export { default as fonts } from './fonts.theme';
export { default as sizes } from './sizes.theme';
export { default as variables } from './variables.theme';

export { NormalizeCSS, GlobalCSS } from './GlobalStyles';
