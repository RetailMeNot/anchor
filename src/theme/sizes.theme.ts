// VARIABLES
import variables from './variables.theme';

const { baseFontSize: base } = variables;

interface Sizes {
  xs: string;
  sm: string;
  md: string;
  lg: string;
  xl: string;
}

const sizes: Sizes = {
  xs: `${base / 4}px`,
  sm: `${base / 2}px`,
  md: `${base * 1}px`,
  lg: `${base * 2}px`,
  xl: `${base * 3}px`,
};

export default {
  padding: { ...sizes },
  margin: { ...sizes },
  border: {
    radius: {
      base: '3px',
    },
    width: {
      base: '1px',
    },
    style: {
      base: 'solid',
    },
  },
};
