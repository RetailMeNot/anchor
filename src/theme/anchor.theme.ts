// ANCHOR THEME
import { radii } from './sizes.theme';
import { MODAL_THEME_KEY, MODAL_THEME } from '../Modal';

// Copied from Adapter.StandardBreakpoints mins
// todo: dedup- they should use a single source
const breakpoints = {
    sm: '576px',
    md: '768px',
    lg: '992px',
    xl: '1200px',
};

interface AnchorTheme {
    colors?: any;
    breakpoints?: any;
    radii?: any;
    [key: string]: any;
}

export const anchorTheme: AnchorTheme = {
    breakpoints,
    radii,
    [MODAL_THEME_KEY]: MODAL_THEME,
};
