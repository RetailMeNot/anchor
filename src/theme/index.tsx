// Theme Files
import { typography } from './typography.theme';
import { ColorsTheme } from './colors.theme';
import { ThemeProvider as XstyledThemeProvider } from '@xstyled/styled-components';
import styled from '@xstyled/styled-components';
import { space as spaceStyles, SpaceProps } from '@xstyled/system';
import { BUTTON_KEY, BUTTON_THEME } from '../Button';
import { INPUT_KEY, INPUT_THEME } from '../Form/Input/utils';
import React from 'react';
import {Button} from '../Button';
import classNames from 'classnames';

export const RootTheme = {
    typography,
    radii: { none: '0', base: '4px', modal: '8px', circular: '1000px' },
    [BUTTON_KEY]: BUTTON_THEME,
    [INPUT_KEY]: INPUT_THEME,
    colors: ColorsTheme,
    fonts: {
        base:
            'Avenir Next, Segoe UI, Helvetica Neue, Helvetica, Roboto, sans-serif',
        mono: `'SFMono-Regular', Consolas, Liberation Mono, Menlo, Courier, monospace`,
    },
    // These are all minimums
    breakpoints: {
        xs: 0,
        sm: 576,
        md: 768,
        lg: 992,
        xl: 1200,
        xxl: 1600,
    },
    // todo: we should likely remove the color.borders colors and just use this
    // or find a way to reference the colors.border colors here
    // e.g. `thin solid borders.base`
    borders: {
        light: `thin solid ${ColorsTheme.borders.light}`,
        base: `thin solid ${ColorsTheme.borders.base}`,
        dark: `thin solid ${ColorsTheme.borders.dark}`,
    },

    skeleton: {
        colorStart: '#E7E7E7',
        colorEnd: '#D3D3D3',
    },
};

// Exporting out @xstyled's ThemeProvider. Only way I could get it to work w/o issues.
// export class ThemeProvider2 extends XstyledThemeProvider {}

interface IterableObject {
    [s: number]: number | string | boolean | IterableObject;
  }

interface DebugProps {
    property: number | string | boolean | IterableObject;
    propertyName: string;
    rootProperty?: boolean;
}

interface StyledDebugProps {
    indent?: number;
}

const StyledDebug = styled('div')<StyledDebugProps>`
    font-family: mono;
    margin-top: 0.5rem;
    color: white;

    .category .category h2 {
        background-color: #ec8947;
    }

    .category .category .box {
        border-color: #ec8947;
    }

    .category .category .category h2 {
        background-color: #b84343;
        color: white;
    }

    .category .category .category .box {
        border-color: #b84343;
    }

    .category .category .category .category h2 {
        background-color: #91cdd2;
        color: black;
    }

    .category .category .category .category .box {
        border-color: #91cdd2;
    }

    .category .category .category .category .category h2 {
        background-color: #567224;
        color: white;
    }

    .category .category .category .category .category .box {
        border-color: #567224;
    }
`;

const StyledH2 = styled('h2')`
    ${spaceStyles}
    background-color: #ffea92;
    color: black;
    margin-bottom: 0;
    margin-top: 0;
    padding: 0.25rem 0.5rem 0.25rem 1rem;
    font-size: 1rem;
    cursor: pointer;
    border-radius: 0 1rem 1rem 0;
`;

interface BoxProps extends SpaceProps {
    show?: boolean;
}

const Box = styled('div')<BoxProps>`
    ${spaceStyles}
    width: 20rem;
    border-left: solid #ffea92 2px;
    margin: 0.5rem;
    ${({show}) => `display: ${show ? 'block' : 'none'}`}
`;

const Category = styled('div')<SpaceProps>`
    ${spaceStyles}
    width: 20rem;
`;


const ColorBox = styled('div')<any>`
    width: 1rem;
    height: 1rem;
    display: inline-block;
    vertical-align: middle;
    margin-left: 0.5rem;
    background-color: ${({backgroundColor}) => backgroundColor};
    border: solid white 1px;
`;

const Debug = ({property, propertyName}: DebugProps) => {
    const [show, setShow] = React.useState(false);

    return(
        <StyledDebug>
            {(property || property === 0)
                ? (
                    typeof property === 'number' ||
                    typeof property === 'string' ||
                    typeof property === 'boolean' ||
                    typeof property === 'function'
                        ? (
                                <>
                                    <span style={{fontWeight: 'bold'}}>{propertyName}: </span>
                                    <span>
                                        {typeof property === 'function'
                                            ? 'function'
                                            : /^#[a-zA-Z0-9]{4,7}/.test(property.toString())
                                                ? (<>
                                                    <span style={{fontStyle: 'italic'}}>{property.toString()}</span>
                                                    <ColorBox backgroundColor={property} />
                                                    </>)
                                                : <span style={{fontStyle: 'italic'}}>{property.toString()}</span>
                                        }
                                    </span>
                                </>
                            )
                        :
                            <Category pl={0} className="category">
                                <StyledH2 onClick={() => setShow(!show)}>{propertyName}</StyledH2>
                                <Box pl={2} className="box" show={show}>
                                    {Object.values(property).map((_property, index) => (
                                        <Debug
                                            key={index}
                                            property={_property}
                                            propertyName={Object.getOwnPropertyNames(property)[index]}
                                        />
                                    ))}
                                </Box>
                            </Category>
                    )
                : (
                    <>
                       Empty.
                    </>
                )
            }
        </StyledDebug>
    );
};

const DebugWrapper = styled('div')`
    background-color: rgba(0,0,0,.75);
    border-left-color: transparent;
    position: fixed;
    left: 0;
    top: 0;
    height: 100%;
    overflow: hidden;
    width: 30rem;
    padding: 1rem 2rem 1rem 0;

    &.closed {
        height:2.4rem;
        width: auto;
        padding-right: 1rem;

        .debug-scroll {
            display: none;
        }
    }
`;

const DebugScroll = styled('div')`
    height: 100%;
    overflow-y: auto;
    width: 100%;
`;

interface ThemeProviderProps {
    children?: any;
    debug?: boolean;
    theme?: any;
}

export const ThemeProvider = ({ children, debug, theme }: ThemeProviderProps) => {
    const isOpen = localStorage.themeProviderDebug === 'true';
    const [isOpenState, setIsOpenState] = React.useState(isOpen);

    return (
        <XstyledThemeProvider theme={theme}>
            {debug &&
                <DebugWrapper className={classNames(!isOpenState ? 'closed' : null)}>
                    <Button
                        onClick={() => {
                            localStorage.themeProviderDebug = !isOpenState;
                            setIsOpenState(!isOpenState);
                        }}
                        size="sm"
                    >
                        ThemeProvider Debug
                    </Button>

                    <DebugScroll className="debug-scroll">
                        <Debug property={theme} propertyName="Theme" />
                    </DebugScroll>
                </DebugWrapper>
            }
            {children}
        </XstyledThemeProvider>
    );
};
