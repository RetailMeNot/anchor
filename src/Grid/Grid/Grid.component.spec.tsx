// REACT
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { ThemeProvider } from '@xstyled/styled-components';
// COMPONENT
import { Grid, Cell } from '../index';
import {
    GridContext,
    createResponsiveObject,
    getResponsiveValue,
    sortBreakpoints,
} from '../utils';
import { RootTheme } from '../../theme';
// ENZYME
import { shallow, mount } from 'enzyme';

// TEST SETUP
const subject = (
    <ThemeProvider theme={RootTheme}>
        <Grid>
            <Cell>Cell 1</Cell>
            <Cell>Cell 2</Cell>
        </Grid>
    </ThemeProvider>
);
const wrapper = mount(subject);
const component = shallow(subject);

describe('Component: Grid & Cell', () => {
    it('should be defined', () => {
        expect(subject).toBeDefined();
        expect(wrapper).toBeDefined();
        expect(component).toBeDefined();
        const tree = renderer.create(subject).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should hide a cell when set to 0 width', () => {
        const testSubject = (
            <Grid>
                <Cell width={0}>Hidden</Cell>
            </Grid>
        );
        const test = mount(testSubject);

        expect(test.find('.anchor-cell.hide').exists()).toBeTruthy();
    });
});

describe('Component: Cell', () => {
    it('should render a number for width if one is supplied', () => {
        const Mock = ({ children }: any) => (
            <GridContext.Provider
                value={{
                    innerWidth: 1024,
                    breakpoints: sortBreakpoints(RootTheme.breakpoints),
                    debug: false,
            }}>
                {children}
            </GridContext.Provider>
        );

        const subject = (<Mock><Cell width={1} left={1} top={1} /></Mock>);
        const component = mount(subject);

        console.log(component.debug());

        // expect(component.find('[width=1]')).toHaveLength(1);
        expect(true);
    });
});

describe('Utilities: Grid & Cell', () => {
    const testObject = { xs: 0, md: 900, sm: 600 };
    const responsiveSettings = { xs: 12, md: 6 };
    const innerWidth = 780;
    const sortedBreakpoints = sortBreakpoints(testObject);
    const responsiveObject = createResponsiveObject(
        responsiveSettings,
        sortedBreakpoints
    );

    it('sortBreakpoints should return a sorted array from an object', () => {
        const lastBreakpoint = sortedBreakpoints[sortedBreakpoints.length - 1];
        const value = Object.values(lastBreakpoint)[0];

        expect(value).toEqual(900);
    });

    it('createResponsiveObject should return a mobile first object based on supplied breakpoints', () => {
        const key = Object.keys(responsiveObject)[1];
        const value = Object.values(responsiveObject)[1];

        expect(key).toEqual('sm');
        expect(value).toEqual(12);
    });

    it('getResponsiveValue should return the value associated from a breakpoint based on window width.', () => {
        const value = getResponsiveValue(
            responsiveObject,
            innerWidth,
            sortedBreakpoints
        );

        expect(value).toEqual(12);
    });
});
