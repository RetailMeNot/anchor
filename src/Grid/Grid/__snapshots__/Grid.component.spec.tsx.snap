// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Grid & Cell should be defined 1`] = `
.c1 {
  height: 100%;
  min-width: 0;
  grid-column-end: span 1;
  grid-row-end: span 1;
  background-color: rgba(255,0,0,0.4);
}

.c0 {
  display: grid;
  height: auto;
  grid-auto-flow: row;
  grid-auto-rows: minmax(1.25rem,auto);
  grid-template-rows: repeat(3,1fr);
  grid-template-columns: repeat(12,1fr);
  grid-gap: 0.5rem;
  -webkit-column-gap: 5rem;
  column-gap: 5rem;
  row-gap: 1rem;
  grid-template-areas: 'head head' 'nav main' 'nav foot';
  -webkit-box-pack: initial;
  -webkit-justify-content: initial;
  -ms-flex-pack: initial;
  justify-content: initial;
  background-color: rgba(255,0,0,0.4);
}

<div
  className="c0 anchor-grid"
  rows={3}
>
  <div
    className="c1 anchor-cell"
    height={1}
    width={1}
  >
    Cell 1
  </div>
  <div
    className="c1 anchor-cell"
    height={1}
    width={1}
  >
    Cell 2
  </div>
</div>
`;
